Add "direct graphics" mode for Zebra ZPL printers.

blueshiftlabs opened the pull request:
This mode uses the ^GF ZPL command to directly emit graphics into the
printer's raster buffer, rather than storing the page raster into an
on-device ZPL graphic, then emitting that graphic into the buffer later.
Some printers (at least the ZP888/ZD120, possibly others) do not have
sufficient onboard memory to store and recall an entire label's worth of
graphics, causing prints on those devices to fail using the existing
filter driver. This "direct mode" driver works on such printers.

zdohnal commented:
Hi Justin,

thank you for the PR!

The intention to support a new mode in CUPS seems fine, however it concerns printer drivers, which are deprecated in CUPS for long time and its support will be removed in the near future.
As for common label printers which don't support any driverless standard there is the [LPrint](https://github.com/michaelrsweet/lprint) project, so it would be great if you adjusted and sent the feature for LPrint as well (in case the mode makes sense to add in LPrint).

Ad PR itself - please check DEVELOPING.md - the PR contains some changes which are only because of different spacing, so it would be great if you cleaned up those. I haven't spotted any 'obvious to me' C issues.

@michaelrsweet Mike, would you mind reviewing the PR from technical point of view once you have a time?

zdohnal commented on data/label.h:
```diff
@@ -8,11 +8,12 @@
  */
 
-#define DYMO_3x0	0		/* Dymo Labelwriter 300/330/330 Turbo */
+#define DYMO_3x0	 0		 /* Dymo Labelwriter 300/330/330 Turbo */
```
Whitespace issue.

zdohnal commented on data/label.h:
```diff
@@ -8,11 +8,12 @@
-#define ZEBRA_EPL_PAGE  0x11            /* Zebra EPL page mode printers */
-#define ZEBRA_ZPL       0x12            /* Zebra ZPL-based printers */
-#define ZEBRA_CPCL      0x13            /* Zebra CPCL-based printers */
+#define ZEBRA_EPL_LINE   0x10            /* Zebra EPL line mode printers */
+#define ZEBRA_EPL_PAGE   0x11            /* Zebra EPL page mode printers */
+#define ZEBRA_ZPL        0x12            /* Zebra ZPL-based printers */
+#define ZEBRA_CPCL       0x13            /* Zebra CPCL-based printers */
```
Whitespace issues.

zdohnal commented on data/label.h:
```diff
@@ -8,11 +8,12 @@
+#define ZEBRA_ZPL_DIRECT 0x14            /* Zebra ZPL-based printers, direct graphics mode */
 
-#define INTELLITECH_PCL	0x20		/* Intellitech PCL-based printers */
+#define INTELLITECH_PCL	 0x20		 /* Intellitech PCL-based printers */
```
Whitespace issue.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -43,14 +43,15 @@
  */
 
-#define DYMO_3x0	0		/* DYMO Labelwriter 300/330/330 Turbo */
+#define DYMO_3x0	 0		/* DYMO Labelwriter 300/330/330 Turbo */
```
Whitespace issue.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -43,14 +43,15 @@
-#define ZEBRA_EPL_PAGE	0x11		/* Zebra EPL page mode printers */
-#define ZEBRA_ZPL	0x12		/* Zebra ZPL-based printers */
-#define ZEBRA_CPCL	0x13		/* Zebra CPCL-based printers */
+#define ZEBRA_EPL_LINE	 0x10		/* Zebra EPL line mode printers */
+#define ZEBRA_EPL_PAGE	 0x11		/* Zebra EPL page mode printers */
+#define ZEBRA_ZPL	 0x12		/* Zebra ZPL-based printers */
+#define ZEBRA_CPCL	 0x13		/* Zebra CPCL-based printers */
```
Whitespace issues.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -43,14 +43,15 @@
+#define ZEBRA_ZPL_DIRECT 0x14		/* Zebra ZPL-based printers - direct graphics mode */ 
 
-#define INTELLITECH_PCL	0x20		/* Intellitech PCL-based printers */
+#define INTELLITECH_PCL	 0x20		/* Intellitech PCL-based printers */
```
Whitespace issue.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -292,6 +297,39 @@ StartPage(ppd_file_t         *ppd,	/* I - PPD file */
 
+    case ZEBRA_ZPL_DIRECT :
+       /*
+	* Start label...
+	*/
```
Incorrect spacing here and further down in the file.

blueshiftlabs replied:
This looks accurate to me? The file as it stands is a bastardized mix of spaces and tabs. I've tried to match surroundings as best as I could.

michaelrsweet reviewed (changes requested):
OK, so from a technical perspective I am hesitant to change anything in the CUPS rastertolabel driver, both because drivers are deprecated and because I've moved development of this driver over to LPrint.

Some specific feedback:

1. This driver has been in use for 20+ years, and this is the first time we've had anyone report issues with memory limits like this. The memory requirements are not high (128k to support 4x6 labels at 203 dpi) and ZPL supports vector graphics.
3. Knowing whether to use ZPL normal or ZPL direct printing mode is not something we want the user to have to know. If the direct mode works with all printers then my preferred solution would be to switch to it instead (more generic, better compatibility across the board)
4. IIRC (and I'll need to dig out my ZPL programming manual to verify) the direct graphics mode doesn't support copies, so we'd either need to change the PPDs to specify manual copy generation or buffer the page in memory in rastertolabel.
5. *If* we take this change, I think it should wait for CUPS 2.5. In the meantime I can also look into using this for LPrint as well.

Regardless, I definitely want to test this extensively with the Zebra printers in my office. Unfortunately, I won't be back to my office until September...

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,144 +530,75 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
-	if (header->NumCopies > 1)
-	  printf("^PQ%d, 0, 0, N\n", header->NumCopies);
+	puts("^XA");
+        puts("^IDR:CUPS.GRF^FS");
+	puts("^XZ");
```
Incorrect spacing.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,144 +530,75 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
 
        /*
-        * Display the label image...
+        * Cut the label as needed...
```
According developing README every 8 spaces have to be 1 tab. Please apply the same on other places.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,144 +530,75 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
 
        /*
-        * Set print width...
+        * Display the label image...
```
Spaces -> tab.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -292,6 +297,39 @@ StartPage(ppd_file_t         *ppd,	/* I - PPD file */
 	LastSet    = 0;
         break;
 
+    case ZEBRA_ZPL_DIRECT :
+       /*
+        * Start label...
+        */
+
+        puts("^XA");
+
+       /*
+        * Emit label header...
+        */
+
+        ZPLEmitLabelHeader(ppd, header);
+
+       /*
+        * Set darkness...
+        */
+
+        if (header->cupsCompression > 0 && header->cupsCompression <= 100)
+          printf("~SD%02u\n", 30 * header->cupsCompression / 100);
+
+       /*
+        * Begin graphics format...
+        */
+
+        puts("^FO0,0");
+        printf("^GFB,%d,%d,%d,",
+               header->cupsHeight * header->cupsBytesPerLine,
+               header->cupsHeight * header->cupsBytesPerLine,
+               header->cupsBytesPerLine);
+
+        break;
+
+
```
Incorrect indentation, check DEVELOPING.md.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,144 +530,75 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
 
        /*
-        * Set print rate...
+        * End the label and eject...
```
Incorrect indentation.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,144 +530,75 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
 
        /*
-        * Put label home in default position (0,0)...
+        * Delete the label image...
```
incorrect indentation.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,144 +530,75 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
 	*/
 
-	puts("^FO0,0^XGR:CUPS.GRF,1,1^FS");
+      	if (header->CutMedia)
```
Incorrect indentation.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,144 +530,75 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
-        * End the label and eject...
-	*/
+    case ZEBRA_ZPL_DIRECT :
+        if (Canceled)
+        {
+         /*
+          * End graphics download by writing enough newline characters
+          * to finish the download.
+          *
+          * We don't know how far into the download we are, so just send
+          * an entire graphic's worth of newlines. (The excess newlines
+          * will be ignored by the printer.)
+          */
+          val = header->cupsHeight * header->cupsBytesPerLine;
+          while (val-- > 0)
+          {
+            putchar('\n');
+          }
 
-	puts("^XZ");
+         /*
+          * Cancel current partially-complete format.
+          */
+          puts("~JX");
+          break;
+        }
```
Incorrect indentation - every 8 spaces must be converted to 1 tab.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,144 +530,75 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
 
        /*
-        * Delete the label image...
+        * End label...
```
Incorrect indenting.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,144 +530,75 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
-	puts("^XA");
-        puts("^IDR:CUPS.GRF^FS");
-	puts("^XZ");
+        puts("^XZ");
```
Incorrect indenting.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,144 +530,75 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
 
-      	if (header->CutMedia)
-	  puts("^CN1");
+        if (header->CutMedia)
+          puts("^CN1");
```
Incorrect indenting.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -870,6 +839,14 @@ OutputLine(ppd_file_t         *ppd,	/* I - PPD file */
         break;
 
+    case ZEBRA_ZPL_DIRECT :
+      /*
+       * We're already in the ^GF command, so just emit the graphics data.
+       */
+       fwrite(Buffer, 1, header->cupsBytesPerLine, stdout);
+       fflush(stdout);
+       break;
```
Incorrect indentation.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -1079,6 +1056,130 @@ ZPLCompress(unsigned char repeat_char,	/* I - Character to repeat */
+ZPLEmitLabelHeader(ppd_file_t *ppd,             /* I - PPD file */
+                   cups_page_header2_t *header) /* I - Page header */
+{
+  int          val;
+  ppd_choice_t *choice;
```
Use tabs between type and name.

zdohnal reviewed (changes requested):
There are other code style errors - would you mind fixing them according the developing readme? I know some parts of the original code doesn't match the standard, it will be fixed in the future as well.

blueshiftlabs commented:
> OK, so from a technical perspective I am hesitant to change anything in the CUPS rastertolabel driver, both because drivers are deprecated and because I've moved development of this driver over to LPrint.
> 
> Some specific feedback:
> 
> 1. This driver has been in use for 20+ years, and this is the first time we've had anyone report issues with memory limits like this. The memory requirements are not high (128k to support 4x6 labels at 203 dpi) and ZPL supports vector graphics.

Memory issues were only my hypothesis here. After a bit more investigation, it seems that the ZP888 only supports a very small subset of ZPL. This doesn't surprise me - the ZP888 is designed as a super-cheap Zebra desktop printer for the Chinese market, with a sales price of around $100.

Even catting out a super-basic "Hello, world!" ZPL script directly to `/dev/usb/lp1` does nothing, so it's entirely possible that the ZP888 doesn't support vector graphics or any of the "fancy" parts of ZPL at all. I went down this path because I sniffed how the printer worked under Windows, and it was using ^GF instead.

The exact script I sent to the printer:

```
^XA
^FO50,50^ADN,36,20^FDHello, world!^FS
^XZ
```

All that happens when I send it is that the light on it blinks, but no label comes out.

> 2. Knowing whether to use ZPL normal or ZPL direct printing mode is not something we want the user to have to know. If the direct mode works with all printers then my preferred solution would be to switch to it instead (more generic, better compatibility across the board)

I'll leave that decision to you - I split this out because I thought that tossing the driver code that's worked just fine for, as you say, 20+ years would be a bad idea considering I only have one Zebra printer to test with, and said printer seems to be an oddball in Zebra's line. This _does_ seem to be a more compatible way of dealing with raster graphics, but there's some things that don't work (^PQ for copies, ^PO to have the label print bottom-edge first) that'll need to be dealt with elsewhere in the stack.

> 3. IIRC (and I'll need to dig out my ZPL programming manual to verify) the direct graphics mode doesn't support copies, so we'd either need to change the PPDs to specify manual copy generation or buffer the page in memory in rastertolabel.

I tested this and you're correct.

> 4. _If_ we take this change, I think it should wait for CUPS 2.5. In the meantime I can also look into using this for LPrint as well.
> 

I could take a cut at trying to port this logic to LPrint, although I've never used LPrint before, so it may take me a bit, and I can't guarantee my PR won't break other things there.

> Regardless, I definitely want to test this extensively with the Zebra printers in my office. Unfortunately, I won't be back to my office until September...

No rush on this - I solved my immediate problem and I'm fine using a locally-patched version until this lands upstream. I just figured I couldn't be the _only_ person to run into this.

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,138 +529,69 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
-	*/
+	if (header->CutMedia)
+	  puts("^CN1");
+        break;
```
Spaces -> tab

zdohnal commented on filter/rastertolabel.c:
```diff
@@ -492,138 +529,69 @@ EndPage(ppd_file_t          *ppd,	/* I - PPD file */
 
        /*
-        * Set print rate...
+        * Display the label image...
```
Spaces -> tab

zdohnal reviewed (changes requested):
The last two spaces -> tab issues.

michaelrsweet commented:
One of my Zebra printers supports this, one doesn't... Not for CUPS 2.x, might consider as an enhancement for LPrint if we can auto-detect support.
